File: src/java/com/springbok/center/SatelliteCatalogTest.java
package com.springbok.center;

import static org.junit.Assert.*;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.ArrayList;
import java.util.zip.DataFormatException;

import org.junit.After;
import org.junit.Before;
import org.junit.Test;

import com.springbok.twobody.EarthConstants;
import com.springbok.twobody.ModJulianDate;
import com.springbok.utility.TestUtility;

public class SatelliteCatalogTest {

	// Directory containing catalog file
	private final String catDir_input = "dat/java/com/springbok/center";
	// File containing two line element sets
	private final String catFNm_input = "tles.txt";

	// Array list containing two line element set parameters constructed from
	// the file
	private ArrayList<SatelliteCatalog.TwoLineElementSet> catalog_input;

	@Before
	public void setUp() throws Exception {

		catalog_input = new ArrayList<SatelliteCatalog.TwoLineElementSet>();

		// this.doyConverter = new DayOfYearConversions();

		String lineOne = "1 00005U 58002B   14224.13195843 -.00000025  00000-0 -42007-4 0  2491";

		// Line 1
		// Columns... Example.......... Description
		// ------------------------------------------------------------------------
		// 3-7....... 25544............ Object Identification Number
		long objectId = 5;
		// 8......... U................ Elset Classification
		String classification = "U";
		// 10-17..... 98067A........... International Designator
		String intlDesignator = "58002B  ";
		// 19-32..... 04236.56031392... Element Set Epoch
		ModJulianDate epoch = ModJulianDate.convertDayOfYear(2014, 224.13195843);
		// 34-43..... .00020137........ First Time Derivative of Mean Motion
		double nDot = -.00000025;
		// 45-52..... 00000-0.......... Second Time Derivative of Mean Motion
		// ............................ (decimal point assumed)
		double nDotDot = 0.0;
		// 54-61..... 16538-3.......... B* Drag Term
		double bStar = -0.42007e-4;
		// 63........ 0................ Element Set Type
		long elSetType = 0;
		// 65-68..... 513.............. Element Number
		long elementNum = 249;
		// 69........ 5................ Checksum

		String lineTwo = "2 00005 032.4486 331.6386 1848534 087.8835 292.2681 18.04384562973062";

		// Line 2
		// Columns... Example.......... Description
		// ------------------------------------------------------------------------
		// 9-16...... 51.6335.......... Orbit Inclination (degrees)
		double i = 032.4486 * Math.PI / 180.0;
		// 18-25..... 344.7760......... Right Ascension of Ascending Node
		// ............................ (degrees)
		double Omega = 331.6386 * Math.PI / 180.0;
		// 27-33..... 0007976.......... Eccentricity
		// ............................ (decimal point assumed)
		double e = 0.1848534;
		// 35-42..... 126.2523......... Argument of Perigee (degrees)
		double omega = 87.8835 * Math.PI / 180.0;
		// 44-51..... 325.9359......... Mean Anomaly (degrees)
		double M = 292.2681 * Math.PI / 180.0;
		// 53-63..... 15.70406856...... Mean Motion (revolutions/day)
		double n = 18.04384562 * 2.0 * Math.PI / 86400;
		double a = Math.pow(EarthConstants.GM_oplus / Math.pow(n, 2), 1.0 / 3.0) / EarthConstants.R_oplus;
		// 64-68..... 32890............ Revolution Number at Epoch
		long revAtEpoch = 97306;
		// 69........ 3................ Checksum

		SatelliteCatalog.TwoLineElementSet tle = new SatelliteCatalog.TwoLineElementSet(lineOne, lineTwo, objectId, classification, intlDesignator, epoch,
				nDot, nDotDot, bStar, elSetType, elementNum, i, Omega, e, omega, M, n, a, revAtEpoch);

		this.catalog_input.add(tle);

		lineOne = "1 40114U 14047F   14228.17412853 -.00000069  00000-0  00000-0 0    69";

		// Line 1
		// Columns... Example.......... Description
		// ------------------------------------------------------------------------
		// 3-7....... 25544............ Object Identification Number
		objectId = 40114;
		// 8......... U................ Elset Classification
		classification = "U";
		// 10-17..... 98067A........... International Designator
		intlDesignator = "14047F  ";
		// 19-32..... 04236.56031392... Element Set Epoch
		epoch = ModJulianDate.convertDayOfYear(2014, 228.17412853);
		// 34-43..... .00020137........ First Time Derivative of Mean Motion
		nDot = -.00000069;
		// 45-52..... 00000-0.......... Second Time Derivative of Mean Motion
		// ............................ (decimal point assumed)
		nDotDot = 0.0;
		// 54-61..... 16538-3.......... B* Drag Term
		bStar = 0.0;
		// 63........ 0................ Element Set Type
		elSetType = 0;
		// 65-68..... 513.............. Element Number
		elementNum = 6;
		// 69........ 5................ Checksum

		lineTwo = "2 40114 064.3596 351.0056 0140909 028.2938 338.7343 16.39284109    93";

		// Line 2
		// Columns... Example.......... Description
		// ------------------------------------------------------------------------
		// 9-16...... 51.6335.......... Orbit Inclination (degrees)
		i = 64.3596 * Math.PI / 180.0;
		// 18-25..... 344.7760......... Right Ascension of Ascending Node
		// ............................ (degrees)
		Omega = 351.0056 * Math.PI / 180.0;
		// 27-33..... 0007976.......... Eccentricity
		// ............................ (decimal point assumed)
		e = 0.0140909;
		// 35-42..... 126.2523......... Argument of Perigee (degrees)
		omega = 28.2938 * Math.PI / 180.0;
		// 44-51..... 325.9359......... Mean Anomaly (degrees)
		M = 338.7343 * Math.PI / 180.0;
		// 53-63..... 15.70406856...... Mean Motion (revolutions/day)
		n = 16.39284109 * 2.0 * Math.PI / 86400;
		a = Math.pow(EarthConstants.GM_oplus / Math.pow(n, 2), 1.0 / 3.0) / EarthConstants.R_oplus;
		// 64-68..... 32890............ Revolution Number at Epoch
		revAtEpoch = 9;
		// 69........ 3................ Checksum

		SatelliteCatalog.TwoLineElementSet tle2 = new SatelliteCatalog.TwoLineElementSet(lineOne, lineTwo, objectId, classification, intlDesignator, epoch, nDot, nDotDot,
				bStar, elSetType, elementNum, i, Omega, e, omega, M, n, a, revAtEpoch);

		this.catalog_input.add(tle2);
	}

	@After
	public void tearDown() throws Exception {
	}

	@Test
	// Tests readCatalog method. Note that readCatalog invoked private method
	// fixed_window, which is therefore tested implicitly.
	public void test_readCatalog() throws NumberFormatException, IOException, DataFormatException {

		ArrayList<SatelliteCatalog.TwoLineElementSet> catalog_expected = this.catalog_input;

		ArrayList<SatelliteCatalog.TwoLineElementSet> catalog_actual = SatelliteCatalog.readCatalog(this.catDir_input,
				this.catFNm_input);

		int iObj = 0;

		// String lineOne
		assertTrue(catalog_actual.get(iObj).lineOne.equals(catalog_expected.get(0).lineOne));
		// long objectId
		assertTrue(catalog_actual.get(iObj).objectId == catalog_expected.get(0).objectId);
		// String classification
		assertTrue(catalog_actual.get(iObj).classification.equals(catalog_expected.get(0).classification));
		// String intlDesignator
		assertTrue(catalog_actual.get(iObj).intlDesignator.equals(catalog_expected.get(0).intlDesignator));
		// ModJulianDate epoch
		assertTrue(catalog_actual.get(iObj).epoch.equals(catalog_expected.get(0).epoch));
		// double nDot
		assertTrue(catalog_actual.get(iObj).nDot == catalog_expected.get(0).nDot);
		// double nDotDot
		assertTrue(catalog_actual.get(iObj).nDotDot == catalog_expected.get(0).nDotDot);
		// double bStar
		assertTrue(catalog_actual.get(iObj).bStar == catalog_expected.get(0).bStar);
		// long elSetType
		assertTrue(catalog_actual.get(iObj).elSetType == catalog_expected.get(0).elSetType);
		// long elementNum
		assertTrue(catalog_actual.get(iObj).elementNum == catalog_expected.get(0).elementNum);

		// String lineTwo
		assertTrue(catalog_actual.get(iObj).lineTwo.equals(catalog_expected.get(0).lineTwo));
		// double i
		assertTrue(Math.abs(catalog_actual.get(iObj).i - catalog_expected.get(0).i) < TestUtility.HIGH_PRECISION);
		// double Omega
		assertTrue(
				Math.abs(catalog_actual.get(iObj).Omega - catalog_expected.get(0).Omega) < TestUtility.HIGH_PRECISION);
		// double e
		assertTrue(catalog_actual.get(iObj).e == catalog_expected.get(0).e);
		// double omega
		assertTrue(
				Math.abs(catalog_actual.get(iObj).omega - catalog_expected.get(0).omega) < TestUtility.HIGH_PRECISION);
		// double M
		assertTrue(Math.abs(catalog_actual.get(iObj).M - catalog_expected.get(0).M) < TestUtility.HIGH_PRECISION);
		// double n
		assertTrue(Math.abs(catalog_actual.get(iObj).n - catalog_expected.get(0).n) < TestUtility.HIGH_PRECISION);
		// double a
		assertTrue(Math.abs(catalog_actual.get(iObj).a - catalog_expected.get(0).a) < TestUtility.HIGH_PRECISION);
		// long revAtEpoch
		assertTrue(catalog_actual.get(iObj).revAtEpoch == catalog_expected.get(0).revAtEpoch);

		iObj = catalog_actual.size() - 1;

		// String lineOne
		assertTrue(catalog_actual.get(iObj).lineOne.equals(catalog_expected.get(1).lineOne));
		// long objectId
		assertTrue(catalog_actual.get(iObj).objectId == catalog_expected.get(1).objectId);
		// String classification
		assertTrue(catalog_actual.get(iObj).classification.equals(catalog_expected.get(1).classification));
		// String intlDesignator
		assertTrue(catalog_actual.get(iObj).intlDesignator.equals(catalog_expected.get(1).intlDesignator));
		// ModJulianDate epoch
		assertTrue(catalog_actual.get(iObj).epoch.equals(catalog_expected.get(1).epoch));
		// double nDot
		assertTrue(catalog_actual.get(iObj).nDot == catalog_expected.get(1).nDot);
		// double nDotDot
		assertTrue(catalog_actual.get(iObj).nDotDot == catalog_expected.get(1).nDotDot);
		// double bStar
		assertTrue(catalog_actual.get(iObj).bStar == catalog_expected.get(1).bStar);
		// long elSetType
		assertTrue(catalog_actual.get(iObj).elSetType == catalog_expected.get(1).elSetType);
		// long elementNum
		assertTrue(catalog_actual.get(iObj).elementNum == catalog_expected.get(1).elementNum);

		// String lineTwo
		assertTrue(catalog_actual.get(iObj).lineTwo.equals(catalog_expected.get(1).lineTwo));
		// double i
		assertTrue(Math.abs(catalog_actual.get(iObj).i - catalog_expected.get(1).i) < TestUtility.HIGH_PRECISION);
		// double Omega
		assertTrue(
				Math.abs(catalog_actual.get(iObj).Omega - catalog_expected.get(1).Omega) < TestUtility.HIGH_PRECISION);
		// double e
		assertTrue(catalog_actual.get(iObj).e == catalog_expected.get(1).e);
		// double omega
		assertTrue(
				Math.abs(catalog_actual.get(iObj).omega - catalog_expected.get(1).omega) < TestUtility.HIGH_PRECISION);
		// double M
		assertTrue(Math.abs(catalog_actual.get(iObj).M - catalog_expected.get(1).M) < TestUtility.HIGH_PRECISION);
		// double n
		assertTrue(Math.abs(catalog_actual.get(iObj).n - catalog_expected.get(1).n) < TestUtility.HIGH_PRECISION);
		// double a
		assertTrue(Math.abs(catalog_actual.get(iObj).a - catalog_expected.get(1).a) < TestUtility.HIGH_PRECISION);
		// long revAtEpoch
		assertTrue(catalog_actual.get(iObj).revAtEpoch == catalog_expected.get(1).revAtEpoch);
	}

	@Test
	// Tests writeCatalog method.
	public void test_writeCatalog()
			throws NumberFormatException, IOException, DataFormatException, InterruptedException {

		ArrayList<SatelliteCatalog.TwoLineElementSet> catalog_input = SatelliteCatalog.readCatalog(this.catDir_input, this.catFNm_input);

		String catFNm_output = this.catFNm_input.replace(".txt", ".dat");

		SatelliteCatalog.writeCatalog(catalog_input, this.catDir_input, catFNm_output);

		Runtime r = Runtime.getRuntime();
		Process p = r.exec(
				"diff " + this.catDir_input + "/" + this.catFNm_input + " " + this.catDir_input + "/" + catFNm_output);
		p.waitFor();
		BufferedReader b = new BufferedReader(new InputStreamReader(p.getInputStream()));
		String line = b.readLine();

		assertTrue(line == null);
	}
}

File: src/matlab/com/springbok/operator/SatelliteCatalogTest.m
% Copyright (C) 2022 Springbok LLC
% 
% This program is free software: you can redistribute it and/or modify
% it under the terms of the GNU General Public License as published by
% the Free Software Foundation, either version 3 of the License, or (at
% your option) any later version.
% 
% This program is distributed in the hope that it will be useful, but
% WITHOUT ANY WARRANTY; without even the implied warranty of
% MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
% General Public License for more details.
% 
% You should have received a copy of the GNU General Public License
% along with this program.  If not, see <https://www.gnu.org/licenses/>.
% 
classdef SatelliteCatalogTest < TestUtility
% Tests methods of SatelliteCatalog class.
  
  properties (Constant = true)
    
    % Directory containing catalog file
    catDir = 'dat/matlab/com/springbok/operator';
    % File containing two line element sets
    catFNm = 'tles.txt';
    
    % Semi-major axis [er]
    a = 2 * EarthConstants.R_oplus;
    % Eccentricity [-]
    e = 0.001;
    % Inclination [rad]
    i = 0;
    % Right ascension of the ascending node [rad]
    Omega = 0;
    % Argument of perigee [rad]
    omega = 0;
    % Mean anomaly [rad]
    M = 0;
    % Epoch date number
    epoch = datenum('01/01/2000 12:00:00');
    
  end % properties (Constant = true)
  
  properties (Access = private)
    
    % Structure containing Keplerian and Equinoctial orbit
    % parameters constructed from the file
    catalog
    
  end % properties (Access = private)
  
  methods
    
    function this = SatelliteCatalogTest(logFId, testMode)
    % Constructs a SatelliteCatalogTest.
    %
    % Parameters
    %   logFId -Log file identifier
    %   testMode - Test mode, if 'interactive' then beeps and pauses
      
    % Invoke superclass constructor
      if nargin == 0
        superArgs = {};
        
      else
        superArgs{1} = logFId;
        superArgs{2} = testMode;
        
      end % if
      this@TestUtility(superArgs{:});
      
      % Assign properties
      nObj = 2;
      
      this.catalog(nObj).id = [];
      
      this.catalog(nObj).a = [];
      this.catalog(nObj).e = [];
      this.catalog(nObj).i = [];
      this.catalog(nObj).Omega = [];
      this.catalog(nObj).omega = [];
      this.catalog(nObj).M = [];
      % Keplarian elements.
      
      this.catalog(nObj).dNm = [];
      this.catalog(nObj).rcs = [];
      
      this.catalog(nObj).j = [];
      this.catalog(nObj).h = [];
      this.catalog(nObj).k = [];
      this.catalog(nObj).p = [];
      this.catalog(nObj).q = [];
      this.catalog(nObj).lambda = [];
      % Equinoctial elements.
      
      iObj = 1;
      
      this.catalog(iObj).id = 5;
      n = 18.04384562 * (2 * pi) * (1 / (24 * 60 * 60));
      % [rad/s] = [rev/day] * [rad/rev] * (1 / ([h/d] * [m/h] * [s/m]))
      a = (EarthConstants.GM_oplus / n^2)^(1/3);
      % [km] = [ [km^3/s^2] / [rad/s]^2 ]^(1/3)
      
      this.catalog(iObj).a = a / EarthConstants.R_oplus;
      % [er] = [km] / [km/er]
      this.catalog(iObj).e = 0.1848534;
      % [-]
      this.catalog(iObj).i = (pi / 180) * 32.4486;
      % [rad]
      this.catalog(iObj).Omega = (pi / 180) * 331.6386;
      % [rad]
      this.catalog(iObj).omega = (pi / 180) * 087.8835;
      % [rad]
      this.catalog(iObj).M = (pi / 180) * 292.2681;
      % [rad]
      % Keplarian elements.
      
      this.catalog(iObj).dNm = datenum(2014, 1, 1) - 1 + 224.13195843;
      
      this.catalog(iObj).j = 1;
      this.catalog(iObj).h = ...
          this.catalog(iObj).e * sin(this.catalog(iObj).omega + this.catalog(iObj).Omega);
      this.catalog(iObj).k = ...
          this.catalog(iObj).e * cos(this.catalog(iObj).omega + this.catalog(iObj).Omega);
      this.catalog(iObj).p = tan(this.catalog(iObj).i / 2) * sin(this.catalog(iObj).Omega);
      this.catalog(iObj).q = tan(this.catalog(iObj).i / 2) * cos(this.catalog(iObj).Omega);
      this.catalog(iObj).lambda = ...
          this.catalog(iObj).M + this.catalog(iObj).omega + this.catalog(iObj).Omega;
      % Equinoctial elements.
      
      iObj = 2;
      
      this.catalog(iObj).id = 40114;
      n = 16.39284109 * (2 * pi) * (1 / (24 * 60 * 60));
      % [rad/s] = [rev/day] * [rad/rev] * (1 / ([h/d] * [m/h] * [s/m]))
      a = (EarthConstants.GM_oplus / n^2)^(1/3);
      % [km] = [ [km^3/s^2] / [rad/s]^2 ]^(1/3)
      
      this.catalog(iObj).a = a / EarthConstants.R_oplus;
      % [er] = [km] / [km/er]
      this.catalog(iObj).e = 0.0140909;
      % [-]
      this.catalog(iObj).i = (pi / 180) * 64.3596;
      % [rad]
      this.catalog(iObj).Omega = (pi / 180) * 351.0056;
      % [rad]
      this.catalog(iObj).omega = (pi / 180) * 28.2938;
      % [rad]
      this.catalog(iObj).M = (pi / 180) * 338.7343;
      % [rad]
      % Keplarian elements.
      
      this.catalog(iObj).dNm = datenum(2014, 1, 1) - 1 + 228.17412853;
      
      this.catalog(iObj).j = 1;
      this.catalog(iObj).h = ...
          this.catalog(iObj).e * sin(this.catalog(iObj).omega + this.catalog(iObj).Omega);
      this.catalog(iObj).k = ...
          this.catalog(iObj).e * cos(this.catalog(iObj).omega + this.catalog(iObj).Omega);
      this.catalog(iObj).p = tan(this.catalog(iObj).i / 2) * sin(this.catalog(iObj).Omega);
      this.catalog(iObj).q = tan(this.catalog(iObj).i / 2) * cos(this.catalog(iObj).Omega);
      this.catalog(iObj).lambda = ...
          this.catalog(iObj).M + this.catalog(iObj).omega + this.catalog(iObj).Omega;
      % Equinoctial elements.
      
    end % SatelliteCatalogTest()
    
    function test_readCatalog(this)
    % Tests readCatalog method. Note that readCatalog invoked
    % private method fixed_window, which is therefore tested
    % implicitly.
      
      warning('off');
      
      catalog_actual = SatelliteCatalog.readCatalog(this.catDir, this.catFNm);
      
      warning('on');
      
      iObj = 1;
      
      t = [];
      
      t = [t; isequal(catalog_actual(iObj).id, this.catalog(iObj).id)];
      
      t = [t; isequal(catalog_actual(iObj).a, this.catalog(iObj).a)];
      t = [t; isequal(catalog_actual(iObj).e, this.catalog(iObj).e)];
      t = [t; isequal(catalog_actual(iObj).i, this.catalog(iObj).i)];
      t = [t; isequal(catalog_actual(iObj).Omega, this.catalog(iObj).Omega)];
      t = [t; isequal(catalog_actual(iObj).omega, this.catalog(iObj).omega)];
      t = [t; isequal(catalog_actual(iObj).M, this.catalog(iObj).M)];
      
      t = [t; isequal(catalog_actual(iObj).dNm, this.catalog(iObj).dNm)];
      
      t = [t; isequal(catalog_actual(iObj).j, this.catalog(iObj).j)];
      t = [t; isequal(catalog_actual(iObj).h, this.catalog(iObj).h)];
      t = [t; isequal(catalog_actual(iObj).k, this.catalog(iObj).k)];
      t = [t; isequal(catalog_actual(iObj).p, this.catalog(iObj).p)];
      t = [t; isequal(catalog_actual(iObj).q, this.catalog(iObj).q)];
      t = [t; isequal(catalog_actual(iObj).lambda, this.catalog(iObj).lambda)];
      
      nObj = length(catalog_actual);
      iObj = 2;
      
      t = [t; isequal(catalog_actual(nObj).id, this.catalog(iObj).id)];
      
      t = [t; isequal(catalog_actual(nObj).a, this.catalog(iObj).a)];
      t = [t; isequal(catalog_actual(nObj).e, this.catalog(iObj).e)];
      t = [t; isequal(catalog_actual(nObj).i, this.catalog(iObj).i)];
      t = [t; isequal(catalog_actual(nObj).Omega, this.catalog(iObj).Omega)];
      t = [t; isequal(catalog_actual(nObj).omega, this.catalog(iObj).omega)];
      t = [t; isequal(catalog_actual(nObj).M, this.catalog(iObj).M)];
      
      t = [t; isequal(catalog_actual(nObj).dNm, this.catalog(iObj).dNm)];
      
      t = [t; isequal(catalog_actual(nObj).j, this.catalog(iObj).j)];
      t = [t; isequal(catalog_actual(nObj).h, this.catalog(iObj).h)];
      t = [t; isequal(catalog_actual(nObj).k, this.catalog(iObj).k)];
      t = [t; isequal(catalog_actual(nObj).p, this.catalog(iObj).p)];
      t = [t; isequal(catalog_actual(nObj).q, this.catalog(iObj).q)];
      t = [t; isequal(catalog_actual(nObj).lambda, this.catalog(iObj).lambda)];
      
      this.assert_true( ...
          'SatelliteCatalog', ...
          'readCatalog', ...
          this.HIGH_PRECISION_DESC, ...
          min(t));
      
    end % test_readCatalog()
    
    function test_writeCatalog(this)
    % Tests writeCatalog method.
      
      warning('off');
      
      [catalog, records] = SatelliteCatalog.readCatalog(this.catDir, this.catFNm);
      
      warning('on');
      
      catFNm_output = strrep(this.catFNm, '.txt', '.dat');
      
      SatelliteCatalog.writeCatalog(catalog, records, this.catDir, catFNm_output);
      
      [status, result] = system( ...
          ['diff ', ...
           this.catDir, '/', this.catFNm, ...
           ' ', ...
           this.catDir, '/', catFNm_output]);
      
      this.assert_true( ...
          'SatelliteCatalog', ...
          'writeCatalog', ...
          this.IS_EQUAL_DESC, ...
          isequal(result, ''));
      
    end % test_writeCatalog()
    
  end % methods
  
  % TODO: Test function writeTLE(objId, lnchId, satOrb, outFId)

end % classdef
